File: backend/app/schemas/ai.py
# Source: Doc 10 §Faz 3 Sprint 3.1 — AI Pydantic şemaları
"""AI analiz ve sohbet şemaları."""

from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field


# ── Enum'lar ──


class AISignalAction(str, Enum):
    BUY = "buy"
    SELL = "sell"
    HOLD = "hold"


class AIConfidence(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"


class ChatRole(str, Enum):
    SYSTEM = "system"
    USER = "user"
    ASSISTANT = "assistant"


# ── Analiz Şemaları ──


class AIAnalysisRequest(BaseModel):
    """Sembol AI analiz isteği."""
    symbol: str = Field(..., description="BIST sembol kodu, ör: THYAO")
    period: str = Field(default="daily", description="Periyod: daily / weekly")
    include_indicators: bool = Field(default=True, description="Teknik göstergeleri dahil et")


class AIIndicatorSummary(BaseModel):
    """Gösterge özet bilgisi."""
    rsi: float | None = None
    macd: float | None = None
    macd_signal: float | None = None
    macd_histogram: float | None = None
    macd_crossover: str | None = None
    stoch_k: float | None = None
    stoch_d: float | None = None
    bb_upper: float | None = None
    bb_middle: float | None = None
    bb_lower: float | None = None
    sma_20: float | None = None
    sma_50: float | None = None
    ema_12: float | None = None
    adx: float | None = None
    volume_ratio: float | None = None
    support_level: float | None = None
    resistance_level: float | None = None
    obv_trend: str | None = None


class AIAnalysisResponse(BaseModel):
    """AI analiz yanıtı."""
    symbol: str
    action: AISignalAction
    confidence: AIConfidence
    summary: str = Field(..., description="AI tarafından üretilmiş Türkçe analiz özeti")
    reasoning: str = Field(..., description="Detaylı gerekçe (Türkçe)")
    key_factors: list[str] = Field(default_factory=list, description="Anahtar karar faktörleri")
    target_price: float | None = Field(default=None, description="Tahmini hedef fiyat")
    stop_loss: float | None = Field(default=None, description="Önerilen zarar durdur seviyesi")
    risk_level: str = Field(default="medium", description="Risk seviyesi: low/medium/high")
    indicators: AIIndicatorSummary | None = None
    model_used: str = Field(default="", description="Kullanılan LLM modeli")
    analyzed_at: datetime = Field(default_factory=datetime.utcnow)


# ── Sohbet (Chat) Şemaları ──


class AIChatMessage(BaseModel):
    """Tek bir sohbet mesajı."""
    role: ChatRole
    content: str


class AIChatRequest(BaseModel):
    """AI sohbet isteği."""
    messages: list[AIChatMessage] = Field(
        ..., min_length=1, description="Sohbet geçmişi"
    )
    symbol: str | None = Field(
        default=None, description="Bağlam sembolü (opsiyonel)"
    )


class AIChatResponse(BaseModel):
    """AI sohbet yanıtı."""
    reply: str
    model_used: str = ""
    usage: dict = Field(default_factory=dict)


# ── Sinyal Şemaları ──


class AISignalResponse(BaseModel):
    """Tek bir AI sinyal."""
    symbol: str
    action: AISignalAction
    confidence: AIConfidence
    reason: str
    score: float = Field(ge=0.0, le=1.0, description="Sinyal skoru 0-1")


class AISignalListResponse(BaseModel):
    """AI sinyal listesi."""
    signals: list[AISignalResponse] = []
    model_used: str = ""
    generated_at: datetime = Field(default_factory=datetime.utcnow)


# ── Model Listesi ──


class AIModelInfo(BaseModel):
    """OpenRouter model bilgisi."""
    id: str
    name: str
    context_length: int | None = None
    pricing: dict = Field(default_factory=dict)


class AIModelListResponse(BaseModel):
    """Kullanılabilir model listesi."""
    models: list[AIModelInfo] = []


# ── Ayarlar ──


class AISettingsRequest(BaseModel):
    """AI ayarları güncelleme."""
    model: str | None = None
    temperature: float | None = Field(default=None, ge=0.0, le=2.0)
    max_tokens: int | None = Field(default=None, ge=100, le=16384)


class AISettingsResponse(BaseModel):
    """Mevcut AI ayarları."""
    model: str
    temperature: float
    max_tokens: int
    base_url: str
    api_key_set: bool = Field(description="API key ayarlanmış mı?")

File: frontend/src/types/ai.ts
/**
 * AI TypeScript tipleri.
 * // Source: Doc 10 §Faz 3 Sprint 3.1 + 3.3 — AI types
 */

// ── Enum'lar ──

export type AISignalAction = "buy" | "sell" | "hold";
export type AIConfidence = "low" | "medium" | "high";
export type ChatRole = "system" | "user" | "assistant";
export type AIExperimentStatus = "pending" | "running" | "completed" | "failed";

// ── Gösterge Özeti ──

export interface AIIndicatorSummary {
  rsi: number | null;
  macd: number | null;
  macd_signal: number | null;
  macd_histogram: number | null;
  macd_crossover: string | null;
  stoch_k: number | null;
  stoch_d: number | null;
  bb_upper: number | null;
  bb_middle: number | null;
  bb_lower: number | null;
  sma_20: number | null;
  sma_50: number | null;
  ema_12: number | null;
  adx: number | null;
  volume_ratio: number | null;
  support_level: number | null;
  resistance_level: number | null;
  obv_trend: string | null;
}

// ── Analiz ──

export interface AIAnalysisRequest {
  symbol: string;
  period?: string;
  include_indicators?: boolean;
}

export interface AIAnalysisResponse {
  symbol: string;
  action: AISignalAction;
  confidence: AIConfidence;
  summary: string;
  reasoning: string;
  key_factors: string[];
  target_price: number | null;
  stop_loss: number | null;
  risk_level: string;
  indicators: AIIndicatorSummary | null;
  model_used: string;
  analyzed_at: string;
}

// ── Sohbet ──

export interface AIChatMessage {
  role: ChatRole;
  content: string;
}

export interface AIChatRequest {
  messages: AIChatMessage[];
  symbol?: string;
}

export interface AIChatResponse {
  reply: string;
  model_used: string;
  usage: Record<string, unknown>;
}

// ── Sinyal ──

export interface AISignalResponse {
  symbol: string;
  action: AISignalAction;
  confidence: AIConfidence;
  reason: string;
  score: number;
}

export interface AISignalListResponse {
  signals: AISignalResponse[];
  model_used: string;
  generated_at: string;
}

// ── Model ──

export interface AIModelInfo {
  id: string;
  name: string;
  context_length: number | null;
  pricing: Record<string, string>;
}

export interface AIModelListResponse {
  models: AIModelInfo[];
}

// ── Ayarlar ──

export interface AISettingsRequest {
  model?: string;
  temperature?: number;
  max_tokens?: number;
}

export interface AISettingsResponse {
  model: string;
  temperature: number;
  max_tokens: number;
  base_url: string;
  api_key_set: boolean;
}

// ── A/B Test Deneyleri ──

export interface AIExperimentCreate {
  name: string;
  description?: string;
  model_a: string;
  model_b: string;
  symbols: string[];
  config?: Record<string, unknown>;
}

export interface AIExperimentResultResponse {
  id: string;
  experiment_id: string;
  symbol: string;
  model_id: string;
  action: AISignalAction;
  confidence: AIConfidence;
  score: number;
  reasoning: string | null;
  latency_ms: number;
  token_usage: Record<string, unknown>;
  created_at: string;
}

export interface AIExperimentResponse {
  id: string;
  user_id: string;
  name: string;
  description: string | null;
  model_a: string;
  model_b: string;
  symbols: string[];
  status: AIExperimentStatus;
  config: Record<string, unknown>;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
  updated_at: string | null;
  results: AIExperimentResultResponse[];
}

export interface AIExperimentListResponse {
  experiments: AIExperimentResponse[];
  total: number;
}

// ── Performans ──

export interface AIAccuracyMetric {
  total_analyses: number;
  correct_predictions: number;
  accuracy_rate: number;
  buy_accuracy: number;
  sell_accuracy: number;
  hold_accuracy: number;
}

export interface AIModelPerformance {
  model_id: string;
  total_analyses: number;
  avg_latency_ms: number;
  avg_score: number;
  avg_confidence_distribution: Record<string, number>;
  total_tokens_used: number;
  accuracy: AIAccuracyMetric;
  period_start: string | null;
  period_end: string | null;
}

export interface AIModelComparison {
  model_a: AIModelPerformance;
  model_b: AIModelPerformance;
  winner: string | null;
  comparison_notes: string[];
}

export interface AIModelComparisonResponse {
  comparison: AIModelComparison;
  generated_at: string;
}

export interface AIPerformanceSummary {
  models: AIModelPerformance[];
  total_analyses: number;
  overall_accuracy: number;
  period_days: number;
  generated_at: string;
}
